configfile : "../config/config.yaml"
import pandas as pd

samples = pd.read_csv(config["samples"],index_col="sample", sep ='\t')
IDS=[s for s in list(samples.index)]
all_fq = [ID + "_1" for ID in IDS] + [ID + "_2" for ID in IDS]
wildcard_constraints:
    sample= "ERR\d+", #"[^\/]",#
    #folder ="\w+"

include: "rules/bowtie.smk"
include: "rules/samtools.smk"
include: "rules/qc.smk"
include: "rules/msa.smk"
include: "rules/denovo.smk"
include: "rules/artificialref.smk"
include: "rules/screen.smk"
include: "rules/decontamination.smk"

rule all:
    input:
        #"msa/alignment.fasta",
        #"tree/tree.nwk",
        #"variability/variability.txt",
        #"variability/variability.png",
        #"tree/tree.png",
        #"reference/artificial_reference_ERR4082860.fa",
        #expand("../results/references/artificial/{sample}.fa", sample=IDS),
        #expand("../results/references/best_references/{sample}.fasta", sample=IDS),
        #expand("../results/blast/{sample}.tsv", sample=IDS),
        #"../results/references/artificial/ERR4082859.fa"
        expand("../results/fasta/{sample}.fa", sample=IDS),                 #FINAL RULE belongs to rule consensus in samtools
        #expand("../results/denovo_assembly/{sample}/contigs.fasta", sample=IDS),
        #"../results/denovo_assembly/ERR4082859/contigs.fasta",
        #"../results/fastq/decontaminated/ERR4082859_1.fq"
        #expand("../results/references/best_references/{sample}.fasta", sample=IDS),
        "../results/qc/multiqc_report.html"
        #"../results/sam/contaminations/ERR4082863.sam"
        #"../results/sam_contaminations/ERR4082860.sam",
        #"../results/references/contamination/contamination_reference.1.bt2"
        #"../results/sam/contaminations/ERR4082863.sam",
        # "../results/references/contamination/contamination_reference.1.bt2"
        #"../results/references/references.fa.nhr"
        #expand("../results/sam/contaminations/{sample}.sam", sample=IDS)







